actors/builtin/miner/miner_actor.go

```diff
@@ -481,6 +475,160 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
 	return nil
 }
 
+type DisputeWindowedPoStParams struct {
+	Deadline  uint64
+	PoStIndex uint64 // only one is allowed at a time to avoid loading too many sector infos.
+}
+
+func (a Actor) DisputeWindowedPoSt(rt Runtime, params *DisputeWindowedPoStParams) *abi.EmptyValue {
+	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
+	reporter := rt.Caller()
+
+	if params.Deadline >= WPoStPeriodDeadlines {
+		rt.Abortf(exitcode.ErrIllegalArgument, "invalid deadline %d of %d", params.Deadline, WPoStPeriodDeadlines)
+	}
+
+	currEpoch := rt.CurrEpoch()
+
+	// Note: these are going to be slightly inaccurate as time
+	// will have moved on from when the post was actually
+	// submitted.
+	//
+	// However, these are estimates _anyways_.
+	epochReward := requestCurrentEpochBlockReward(rt)
+	pwrTotal := requestCurrentTotalPower(rt)
+
+	toBurn := abi.NewTokenAmount(0)
+	toReward := abi.NewTokenAmount(0)
+	pledgeDelta := abi.NewTokenAmount(0)
+	powerDelta := NewPowerPairZero()
+	var st State
+	rt.StateTransaction(&st, func() {
+		if !deadlineAvailableForOptimisticPoStDispute(st.ProvingPeriodStart, params.Deadline, currEpoch) {
+			rt.Abortf(exitcode.ErrForbidden, "can only dispute window posts during the dispute window (%d epochs after the challenge window closes)", WPoStDisputeWindow)
+		}
+
+		info := getMinerInfo(rt, &st)
+		disputedPower := NewPowerPairZero()
+		store := adt.AsStore(rt)
+
+		// Check proof
+		{
+			// Find the proving period start for the deadline in question.
+			ppStart := st.ProvingPeriodStart
+			if st.CurrentDeadline < params.Deadline {
+				ppStart -= WPoStProvingPeriod
+			}
+			targetDeadline := NewDeadlineInfo(ppStart, params.Deadline, currEpoch)
+
+			// Load the target deadline.
+			deadlinesCurrent, err := st.LoadDeadlines(store)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+
+			dlCurrent, err := deadlinesCurrent.LoadDeadline(store, params.Deadline)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline")
+
+			// Take the post from the snapshot for dispute.
+			// This operation REMOVES the PoSt from the snapshot so
+			// it can't be disputed again. If this method fails,
+			// this operation must be rolled back.
+			partitions, proofs, err := dlCurrent.TakePoStProofs(store, params.PoStIndex)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load proof for dispute")
+
+			// Load the partition info we need for the dispute.
+			disputeInfo, err := dlCurrent.LoadPartitionsForDispute(store, partitions)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partition info for dispute")
+			disputedPower = disputeInfo.DisputedPower
```

Comment:
It's worth noting again that this includes power for sectors that were active at proof submission time but may have since expired, and that this is and must be only used for penalty calculations, not power accounting. Maybe worth a rename to `penalisedPower`.